### PR TITLE
#3628 Add locale parameter for numberFormat

### LINKED ISSUE #3628: Add locale parameter for numberFormat
### Use case

When mapping from BigDecimal / double to String type (maybe like a number with 2 digits after separator), then it choose separator with honor to default program locale.
So, the results is incosistent. On my computer i have a locale with "ru/RU" and it choose ',' as separator, but on remote tests with locale "en/US" it chooses '.'

### Generated Code

Add a locale parameter with conjuction to numberFormat

### Possible workarounds

Workaround - have a separate util method with manually created DecimalFormat.

### MapStruct Version

MapStruct 1.5.5.Final

### PR DESCRIPTION
**Pull request by Obolrom**
Implementation of enhancement #3628

- [x] add Mapping#locale param
- [x] adjust decimal format creation logic
- [x] add test case for other default lang
- [x] support locale param for Mapping#dateFormat
- [x] integrate locale param for MapMapping & IterableMapping annotations
- [x] add documentation for locale params

e.g. 
```java
@Mapping( target = "bigDecimal1", numberFormat = "#0.#E0", locale = "en/US" )
```

Fixes #3628

### CODE REVIEW

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks for your work on this @Obolrom. It is on the right track. However, it still needs some work before it is in a mergeable state. I've left some inline comments, but there are some other more general remarks as well:

* The use of `new Locale` public constructor. This constructor has been deprecated in Java 19 and there is `Locale.forLanguageTag` available since Java 7, let's use that everywhere instead.
* I see that we are only supporting the `numberFormat` in the `@Mapping` for this. However, we also have `dateFormat` in `@Mapping` and we also have the following (if we are adding the `locale` we need to support it everywhere):
  * `MapMapping#keyNumberFormat`
  * `MapMapping#keyDateFormat`
  * `MapMapping#valueNumberFormat`
  * `MapMapping#valueDateFormat`
  * `IterableMapping#numberFormat`
  * `IterableMapping#dateFormat`
* I don't see any tests with a custom locale. The `NumberFormatConversionRuLocaleTest` is identical to `NumberFormatConversionTest`, with the only difference being that a different default locale is used. What we need instead is in the `NumberFormatConversionTest` (which runs using the default `en` locale to have mappers using a custom locale.
* An additional improvement could be that we store the passed `Locale` to the method in a field, but this is more complex and this can be an additional improvement.

File: processor/src/main/resources/org/mapstruct/ap/internal/conversion/CreateDecimalFormat.ftl [hunk 1]
```diff
@@ -5,9 +5,10 @@
     Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 
 -->
-private DecimalFormat ${name}( String numberFormat ) {
+private DecimalFormat ${name}( String numberFormat, Locale locale ) {
 
     DecimalFormat df = new DecimalFormat( numberFormat );
     df.setParseBigDecimal( true );
+    df.setDecimalFormatSymbols( DecimalFormatSymbols.getInstance( locale ) );
     return df;
```
[the remark this answers is not in the document]
Comment (Obolrom) [reply processor/src/main/resources/org/mapstruct/ap/internal/conversion/CreateDecimalFormat.ftl:13]: We can do even better, DecimalFormat has a constructor with additional param (DecimalFormatSymbols symbols).
So we can do this

```java
DecimalFormat df = new DecimalFormat( numberFormat, DecimalFormatSymbols.getInstance( locale ) );
df.setParseBigDecimal( true );
return df;
```

[the remark this answers is not in the document]
Comment (filiphr) [reply processor/src/main/resources/org/mapstruct/ap/internal/conversion/CreateDecimalFormat.ftl:13]: That would indeed be better. I also can't seem to find the constructor I proposed

File: core/src/main/java/org/mapstruct/Mapping.java [hunk 2]
```diff
@@ -188,6 +188,8 @@
      */
     String numberFormat() default "";
 
+    String locale() default "";
```
[the remark this answers is not in the document]
Comment (Obolrom) [reply core/src/main/java/org/mapstruct/Mapping.java:208]: added

**Comment by Obolrom** [comment]
@filiphr I have integrated locale param for IterableMapping & MapMapping, added documentation, fixed tests, and made the code a little cleaner. So it is ready for review.

However, I didn't understand your idea about 
> * An additional improvement could be that we store the passed Locale to the method in a field, but this is more complex and this can be an additional improvement.

**Comment by filiphr** [comment]
>However, I didn't understand your idea about

Ignore that for now. It isn't that important

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/BigDecimalToStringConversion.java [hunk 3]
```diff
@@ -76,6 +76,14 @@ private void appendDecimalFormatter(StringBuilder sb, ConversionContext conversi
...
             sb.append( "\"" );
         }
+        if ( conversionContext.getLocale() != null ) {
+            sb.append( ", java.util.Locale.forLanguageTag( \"" );
+            sb.append( conversionContext.getLocale() );
+            sb.append( "\" )" );
+        }
+        else {
+            sb.append( ", java.util.Locale.getDefault() " );
+        }
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/BigDecimalToStringConversion.java:86]: We strive to have as much readable code as possible. having a full qualified name here isn't readable 😉. Please have a look at how we are doing it in the `getFromExpression` for the `BigDecimal` import

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/BigIntegerToStringConversion.java [hunk 4]
```diff
@@ -84,6 +84,14 @@ private void appendDecimalFormatter(StringBuilder sb, ConversionContext conversi
...
             sb.append( "\"" );
         }
+        if ( conversionContext.getLocale() != null ) {
+            sb.append( ", Locale.forLanguageTag( \"" );
+            sb.append( conversionContext.getLocale() );
+            sb.append( "\" )" );
+        }
+        else {
+            sb.append( ", Locale.getDefault() " );
+        }
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/BigIntegerToStringConversion.java:94]: Same comment as in `BigDecimalToStringConversion`

**Review by filiphr — CHANGES_REQUESTED** [review]
I've missed some things in my previous review. Have a look at my comments. We strive to have a readable generated code. What is being done now is not really readable, let's try to not have full qualified names if possible.

**Comment by Obolrom** [comment]
@filiphr
I have fixed the issues you mentioned in the last comment.

However, right now we have some code duplications
`WrapperToStringConversion::appendDecimalFormatter` is identical to `PrimitiveToStringConversion::appendDecimalFormatter`
and
`BigDecimalToStringConversion::appendDecimalFormatter` is identical to `BigIntegerToStringConversion::appendDecimalFormatter`.

We can make `appendDecimalFormatter` method abstract and move it to `AbstractNumberToStringConversion`.
However, in this case, the code will still be duplicated.
What do you think of that?

**Comment by Obolrom** [comment]
@filiphr
Kind reminder to review PR after changes :smile:

**Comment by filiphr** [comment]
Looks better @Obolrom. However, I see that there is a change in behaviour and now we always pass `Locale.getDefault` to certain places. I would say that we should only pass some locale if there is something set by the user. Otherwise, we keep doing what we were doing before. 

i.e. there should be no need to change anything in existing tests.

**Comment by filiphr** [comment]
If you are busy, I can apply these changes while merging

**Comment by Obolrom** [comment]
@filiphr
It would be great if we close this PR today.
Feel free to apply these changes

**Comment by filiphr** [comment]
@Obolrom it will most definitely not be closed today. I'll apply the changes when I start the merge

**Pull request merged**